### ISSUE TITLE
About problems in predict.py: I tried to predict whether two compounds could form a cocrystal with predict.py, but results are very bad.

### ISSUE THREAD

**Issue by MingLunHan** [Saoge123/ccgnet#2]
About problems in predict.py: I tried to predict whether two compounds could form a cocrystal with predict.py, but results are very bad.
this is my sdf 3d format for a compound：

`

 OpenBabel05042118463D

 17 17  0  0  0  0  0  0  0  0999 V2000
   -1.9173    0.6874    0.0034 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.7014    1.3857    0.0150 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.4874    0.6703    0.0070 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.4593   -0.7218    0.0008 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.7428   -1.4247   -0.0091 C   0  0  0  0  0  0  0  0  0  0  0  0
   -1.9529   -0.7231   -0.0142 C   0  0  0  0  0  0  0  0  0  0  0  0
   -3.2967   -1.3795   -0.0408 C   0  0  0  0  0  0  0  0  0  0  0  0
   -4.3301   -0.7249   -0.0461 O   0  0  0  0  0  0  0  0  0  0  0  0
   -3.3233   -2.7270   -0.0626 O   0  0  0  0  0  0  0  0  0  0  0  0
    1.6235   -1.4382    0.0038 O   0  0  0  0  0  0  0  0  0  0  0  0
    1.7056    1.2887    0.0036 O   0  0  0  0  0  0  0  0  0  0  0  0
   -2.8489    1.2545    0.0063 H   0  0  0  0  0  0  0  0  0  0  0  0
   -0.7112    2.4716    0.0294 H   0  0  0  0  0  0  0  0  0  0  0  0
   -0.6827   -2.5043   -0.0137 H   0  0  0  0  0  0  0  0  0  0  0  0
   -2.4804   -3.2086   -0.0612 H   0  0  0  0  0  0  0  0  0  0  0  0
    2.3380   -0.7700    0.0138 H   0  0  0  0  0  0  0  0  0  0  0  0
    1.5637    2.2521   -0.0092 H   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  2  0  0  0  0
  1  6  1  0  0  0  0
  1 12  1  0  0  0  0
  2  3  1  0  0  0  0
  2 13  1  0  0  0  0
  3  4  2  0  0  0  0
  3 11  1  0  0  0  0
  4  5  1  0  0  0  0
  4 10  1  0  0  0  0
  5  6  2  0  0  0  0
  5 14  1  0  0  0  0
  6  7  1  0  0  0  0
  7  8  2  0  0  0  0
  7  9  1  0  0  0  0
  9 15  1  0  0  0  0
 10 16  1  0  0  0  0
 11 17  1  0  0  0  0
M  END
$$$$
`

and this is my 2d sdf format:

`

 OpenBabel05042116122D

  7  6  0  0  0  0  0  0  0  0999 V2000
   -1.7321    0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -2.5981   -0.5000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -3.4641    0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -3.4641    1.0000    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
   -4.3301   -0.5000    0.0000 O   0  0  0  0  0  0  0  0  0  0  0  0
   -0.8660   -0.5000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.0000   -0.0000    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  1  6  1  0  0  0  0
  2  3  1  0  0  0  0
  3  4  2  0  0  0  0
  3  5  1  0  0  0  0
  6  7  1  0  0  0  0
M  END
$$$$
`

I tried with these two formats, my test set has about 150 samples, and the results predicted by ccgnet is very bad, most samples are tagged as 1 (YES), and only few samples are tagged as 0. This is completely different from our own labels. 

Any ideas? Is our format of files wrong?

**Comment by MingLunHan** [comment]
I also tried some samples in training set, model also gives wrong results.

**Comment by Saoge123** [comment]
Hi
sorry for the late reply.
Because I don't know the situation of your data set, I can not determine the problem. Is it convenient for you to send me 150 test samples? It's good to find what caused the bad performance. It may be the code problem or the difference of the data. My e-mail: 289937611@qq.com

**Comment by Saoge123** [comment]
The input of the model should preferably be an optimized 3D structure.
The original CCGNet depend on RDkit, Openbabel and CCDC Python API to calculate sample features, and the model we uploaded was trained by these features. But for better open source, we removed CCDC Python API. So the features only were computed by RDkit and  Openbabel. It is different with the training. This may lead wrong results when you predict the samples in training set. Later, we will upload the original version used in our work.

**Comment by MingLunHan** [comment]
> Hi
> sorry for the late reply.
> Because I don't know the situation of your data set, I can not determine the problem. Is it convenient for you to send me 150 test samples? It's good to find what caused the bad performance. It may be the code problem or the difference of the data. My e-mail: [289937611@qq.com](mailto:289937611@qq.com)

Thank you very much! I will pack up my test set first and send it to your e-mail, thanks for your sincere reply Again!

**Comment by ArtemS123** [comment]
Dear Saoge123,
Predict.py does not appear to operate properly. I've been attempting to estimate Carbamazepine's propensity for cocrystallization with a variety of coformers using ccgnet-OB and predict.py. All of the sdf-input files were taken from this repository. However, the resulting score values for tried combinations looked to be low and unrealistic. For example, the score value for carbamazepine + 4-aminobenzoic acid (known cocrystal) was calculated to be 0.7. The resultant value for carbamazepine + nicotinamide (known cocrystal) was even worse (-3.4). (see attached file). How can I trust predict.py forecasts for arbitrary carbamazepine combinations if the script has failed to predict known cocrystals? Could you kindly comment on that? Perhaps a revision of predict.py and ccgnet-OB is required. Or description of usage of ccgnet-OB is incomplete and I’m doing something wrong.  Also, could you please upload the "original" version of predict.py that utilizes the CCDC Python API? I'd like to try the published version of ccgnet because I already have the CSD system installed on my desktop. Although the "original" predict.py is described on the main page (https://github.com/Saoge123/ccgnet), the file itself is not available.
Looking forward to hear back from you
[cc_test.xlsx](https://github.com/Saoge123/ccgnet/files/9388476/cc_test.xlsx)

**Comment by Saoge123** [comment]
> Dear Saoge123, Predict.py does not appear to operate properly. I've been attempting to estimate Carbamazepine's propensity for cocrystallization with a variety of coformers using ccgnet-OB and predict.py. All of the sdf-input files were taken from this repository. However, the resulting score values for tried combinations looked to be low and unrealistic. For example, the score value for carbamazepine + 4-aminobenzoic acid (known cocrystal) was calculated to be 0.7. The resultant value for carbamazepine + nicotinamide (known cocrystal) was even worse (-3.4). (see attached file). How can I trust predict.py forecasts for arbitrary carbamazepine combinations if the script has failed to predict known cocrystals? Could you kindly comment on that? Perhaps a revision of predict.py and ccgnet-OB is required. Or description of usage of ccgnet-OB is incomplete and I’m doing something wrong. Also, could you please upload the "original" version of predict.py that utilizes the CCDC Python API? I'd like to try the published version of ccgnet because I already have the CSD system installed on my desktop. Although the "original" predict.py is described on the main page (https://github.com/Saoge123/ccgnet), the file itself is not available. Looking forward to hear back from you [cc_test.xlsx](https://github.com/Saoge123/ccgnet/files/9388476/cc_test.xlsx)

Dear ArtemS123:
    I am sorry for the late reply. predict.py has been uploaded, results of the paper can be reproduced. 
    Here, we judge whether a pair of molecules can form cocrystal by the score of prediction. Therefore, for the prediction results, we generally focus on the score ranking, and the higher the score, the greater the possibility of cocrystal formation. If you want to guide the experiment, generally refer to the molecules with the top scores. However, the predicted label' yes' or' no' has relatively little reference significance.
    Note that, the 3D conformation is important to the performance, so you had better to get the best conformation of coformers before runing predict.py.  If you have any questions, please contact us.

**Comment by ArtemS123** [comment]
Dear Saoge123,
Thank you for responding so quickly. I’ve tested the ccgnet with the predict.py file that you kindly uploaded. For the test, I used combination of the compounds provided in the .\data\Test\Test_Table.tab file and the .\data\Test\coformers\ folder. But still, I was not able to reproduce the results shown in your study, particularly values in Fig.4 (see attached file). For example, the combination of Carbamazepine and glycolic acid received a score of 43.1. In the publication, the figure is -8.34. Similar situation is also seen for Carbamazepine + Malic acid, Carbamazepine+ meglumine and other “negative” combinations. The resulting score is high, equivalent to "real" cocrystals. Predict.py also failed to detect the majority of the unfavorable outcomes for indomethacin (see attached file). What solution would you suggest to address these issues? 
[cc_test.xlsx](https://github.com/Saoge123/ccgnet/files/9389230/cc_test.xlsx)

**Comment by Saoge123** [comment]
Dear ArtemS123:
    Your problem may be that the models that load for predition are in the ccgnet-OB. You can use the models in "./ccgnet/inference_model/CC_Models/" to reproduce results of paper.
[cc_test.xlsx](https://github.com/Saoge123/ccgnet/files/9411589/cc_test.xlsx)

**Comment by MingLunHan** [comment]
Questions have been solved.

**Comment by ArtemS123** [comment]
Dear Saoge123,
Thanks a lot for your support. I just wanted to mention that the problem was resolved after I installed the Open Babel GUI software on my PC. Now, I’ve got the same score values as in the paper. Also, I'd like to point out some small flaws in your article. The letters "c" and "d" in the Fig.4 caption are switched; "c" should correspond to indomethacin, while "d" should correspond to paracetamol. In addition, carbamazepine does form cocrystal with DL-tartaric acid (Compound CID: [875](https://pubchem.ncbi.nlm.nih.gov/compound/875)) (see [doi.org/10.1107/S2053229620000868](https://doi.org/10.1107/S2053229620000868)). Thus, this combination should be marked as false negative.
